public/app/features/dashboard/state/PanelModel.ts

```diff
@@ -269,6 +270,33 @@ export class PanelModel implements DataConfigSource, IPanelModel {
     }
   }
 
+  // Attempts to populate each panel target with a well-formed query definition
+  async ensureWellFormedQueries() {
+    if (this.targets && isArray(this.targets)) {
+      // If we modified any targets, publish one update event
+      let shouldPublish = false;
+      for (let i = 0; i < this.targets.length; i++) {
+        const query = this.targets[i];
+        if (!query.datasource && this.datasource) {
+          query.datasource = this.datasource;
+        }
+        if (Object.keys(query).length === 2) {
+          // If query has only refId and datasource, try to set the default query for its datasource
+          const datasourceInstance = await getDataSourceSrv().get(query.datasource);
+          const defaultQuery = datasourceInstance.getDefaultQuery?.(CoreApp.PanelEditor);
```

Comment:
this is already called from here, https://github.com/grafana/grafana/blob/main/public/app/features/query/components/QueryGroup.tsx#L141 

Do we really need ensureWellFormedQueries?

Reply:
It appears that we do, if I don't call `ensureWellFormedQueries`, my test case fails. This may be because of how the PanelModel's targets are being set. I will investigate a little.

Reply:
Confirmed that `newQuery()` is not called when switching datasources, only when the +Query button is clicked.